3.x Validator::fieldsMatching() and Validator::fieldsNotMatching()

inoas opened the pull request:
Ref: https://github.com/cakephp/cakephp/issues/8618
@dakota Let me know if this fits the bill @ https://github.com/cakephp/cakephp/issues/8618#issuecomment-207818465

inoas committed:
add fieldsMatching and fieldsNotMatching

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,44 @@ public static function compareWith($check, $field, $context)
+     */
+    public static function fieldsNotMatching($fieldOne, $fieldTwo, $context)
+    {
+        return !self::fieldsMatching($fieldOne, $fieldTwo, $context);
```
should be static::

inoas committed:
self -> static

markstory commented:
Implementation looks pretty good. We'll need tests before this can be merged though.

inoas committed:
shorten code

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
         return $context['data'][$field] === $check;
     }
 
+
```
Superfluous white line

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     * @param string $fieldOne The field one to match field two against.
+     * @param string $fieldTwo The field two to match field one against.
+     * @param array $context The validation context.
+     * @return bool
```
Success.

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+    /**
+     * Matches two fields with each other.
+     *
+     * If both fields have exactly the same value this method will return true.
```
Nitpicking from a comma nazi: "value, this"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     *
+     * If both fields have exactly the same value this method will return true.
+     *
+     * @param string $fieldOne The field one to match field two against.
```
"first field" and "second field"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     * If both fields have exactly the same value this method will return true.
+     *
+     * @param string $fieldOne The field one to match field two against.
+     * @param string $fieldTwo The field two to match field one against.
```
"second field" and "first field"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+    /**
+     * Validates if two fields do not match each other.
+     *
+     * If both fields have exactly the same value this method will return false.
```
Comma: "value, this"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     *
+     * If both fields have exactly the same value this method will return false.
+     *
+     * @param string $fieldOne The field one to match field two against.
```
"first field" and "second field"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     * If both fields have exactly the same value this method will return false.
+     *
+     * @param string $fieldOne The field one to match field two against.
+     * @param string $fieldTwo The field two to match field one against.
```
"second field" and "first field"

ravage84 commented on src/Validation/Validation.php:
```diff
@@ -322,6 +322,41 @@ public static function compareWith($check, $field, $context)
+     * @param string $fieldOne The field one to match field two against.
+     * @param string $fieldTwo The field two to match field one against.
+     * @param array $context The validation context.
+     * @return bool
```
Success.

markstory reviewed (changes requested):
Implementation looks pretty good. We'll need tests before this can be merged though.

inoas committed:
Add options to return true if fields given do not exist in context, added option to omit typesafe checking, set the typesafe flag to false for notMatching, added tests

inoas commented:
I have added tests.

However when adding tests I noted that the validations would not work "as expected". If fields did not exist, `fieldsMatching()` would return `false` - `fieldsNotMatching()` would invert that to `true`, which was wrong.

Going further down those issues I have added options to:
- disable field checking in context. By default the fields are assumed to be there (true)
- disable type safe checking. By default the type checking is on for fieldsMatching and off for fieldsNotMatching (this overall should be the most strict mode out of the box).

I hope it is still simple enough. *Blows a stupid simple kiss to @dereuromark*.

inoas committed:
Improved tests

lorenzo commented on src/Validation/Validation.php:
```diff
@@ -323,6 +323,80 @@ public static function compareWith($check, $field, $context)
+     * @param array $options Optional options for the check. Default to ['fieldsExpected' => true, 'typeSafe' => true].
+     * @return bool
+     */
+    public static function fieldsMatching($fieldOne, $fieldTwo, $context, array $options = [])
```
This function should live in `Validator.php`. I see this file as a legacy thing that should be drained out of methods instead of adding more.

`Validator.php` is the place where you can guarantee the `$context` being passed correctly.

ADmad replied:
Also isn't `$context` always passed as last argument by caller?

inoas replied:
@ADmad http://api.cakephp.org/3.0/source-class-Cake.Validation.Validation.html#299
@lorenzo Should it be *moved* to `Validator` or have a proxy method in there instead?

ADmad replied:
@inoas Add a integration test, i.e. a table which uses this rule.

inoas replied:
@ADmad yes about the order I considered that `$context` must be last.

inoas committed:
stickler/cs, docbloc gramma

inoas committed:
docblock

inoas commented:
So what's the plan now?
1. Move to Validator? Add proxy to Validator? 
2. Add test based on validator method?

markstory commented:
>  1. Move to Validator? Add proxy to Validator? 

Sounds like the best option to me.

inoas commented:
@markstory so which, MOVE or ADD PROXY ;)?

markstory commented:
I think putting validation logic that requires the `$context` belongs on `Validator`. Moving this validator there as @lorenzo suggested makes sense to me.

inoas committed:
add proxy to Validatior, created unit tests

inoas committed:
punch stickler

inoas committed:
punch stickler even more

inoas committed:
smoke the peace pike with stickler-ci

inoas commented:
1. I have added proxy methods for now to Validator.php as most of them are just that and still live in Validation.php - even some having context.
2. So I spend like 2 hours trying to figure out how to use assertProxy() correctly. `sameAs()` has no options so it is not a problem, but I could not make the unit tests run. I'd be happy about help/guidance here and/or on IRC.

inoas commented:
This probably only needs some tweaking around the tests, which I could not make working. Could you assists with that?

inoas commented on src/Validation/Validation.php:
```diff
@@ -322,39 +322,78 @@ public static function compareWith($check, $field, $context)
+            return $context['data'][$fieldOne] === $context['data'][$fieldTwo];
+        }
+
+        return $context['data'][$fieldOne] == $context['data'][$fieldTwo];
```
Now this sucks a lot: https://github.com/dereuromark/cakephp-tools/blob/master/src/Utility/Utility.php#L31

However given limited time this is nothing I will work on in near future.

dereuromark replied:
Maybe we can port a type safe comparison into the core Utility, and then simply reuse that, then the option is not needed and this PR would be rather issue to finish up and get merged.
If someone wants to help out here, all help is appreciated.

markstory commented:
Closing until there is time to revisit this and fix up the outstanding issues.
